src/reducers/referralsReducer.js

```diff
@@ -32,30 +53,155 @@ export type ReferralsSendingInviteAction = {|
+  | ReferralsInviteAlreadySentAction
+  | ReferralsStateAction;
+
+export type ReferralsReducerState = {
+  isSendingInvite: boolean,
+  addedContactsToInvite: ReferralContact[],
+  alreadyInvitedContacts: ReferralContact[],
+  hasAllowedToAccessContacts: boolean,
+  sentInvitationsCount: SentInvitationsCount,
+};
 
 export const initialState = {
+  addedContactsToInvite: [],
+  alreadyInvitedContacts: [],
   isSendingInvite: false,
+  hasAllowedToAccessContacts: false,
+  sentInvitationsCount: {
+    count: 0,
+    date: '',
+  },
+};
+
+
+const setContacts = (
+  state: ReferralsReducerState,
+  action: ReferralsSetContactsAction,
+): ReferralsReducerState => {
+  const { payload } = action;
+
+  return {
+    ...state,
+    addedContactsToInvite: payload,
+  };
 };
 
-const ratesReducer = (
+const removeContact = (
+  state: ReferralsReducerState,
+  action: ReferralsRemoveContactAction,
+): ReferralsReducerState => {
+  const { payload } = action;
+  const { addedContactsToInvite } = state;
+
+  return {
+    ...state,
+    addedContactsToInvite: [...addedContactsToInvite.filter((contact) => contact.id !== payload)],
+  };
+};
+
+const setInvitations = (
+  state: ReferralsReducerState,
+  action: ReferralsInviteSentAction,
+): ReferralsReducerState => {
+  const { payload } = action;
+  const { alreadyInvitedContacts: _alreadyInvitedContacts, sentInvitationsCount } = payload;
+  const { alreadyInvitedContacts } = state;
+
+  return {
+    ...state,
+    isSendingInvite: false,
+    addedContactsToInvite: [],
+    alreadyInvitedContacts: [...alreadyInvitedContacts, ..._alreadyInvitedContacts],
+    sentInvitationsCount,
+  };
+};
+
+const setAlreadySentInvites = (
+  state: ReferralsReducerState,
+  action: ReferralsInviteAlreadySentAction,
+): ReferralsReducerState => {
+  const { payload } = action;
+  return { ...state, alreadyInvitedContacts: payload };
+};
+
+const setReferralsState = (
+  state: ReferralsReducerState,
+  action: ReferralsStateAction,
+): ReferralsReducerState => {
+  const { payload } = action;
+  return { ...state, ...payload };
+};
+
+
+export default function referralsReducer(
   state: ReferralsReducerState = initialState,
   action: ReferralsReducerAction,
-): ReferralsReducerState => {
+): ReferralsReducerState {
   switch (action.type) {
     case SENDING_INVITE:
       return { ...state, isSendingInvite: true };
 
+    case SET_REFERRALS_STATE:
+      return setReferralsState(state, action);
+
     case INVITE_SENT:
+      return setInvitations(state, action);
+
+    case REFERRAL_INVITE_ERROR:
       return { ...state, isSendingInvite: false };
 
+    case SET_ALREADY_INVITED_CONTACTS:
+      return setAlreadySentInvites(state, action);
+
+    case SET_CONTACTS_FOR_REFERRAL:
+      return setContacts(state, action);
+
+    case REMOVE_CONTACT_FOR_REFERRAL:
+      return removeContact(state, action);
+
+    case ALLOW_ACCESS_PHONE_CONTACTS:
+      return { ...state, hasAllowedToAccessContacts: true };
```

Comment:
What's this for? It looks like we don't use it.

Reply:
I use it here:
https://github.com/pillarwallet/pillarwallet/pull/1770/files/555baa96dce2e76c67392bda120abc3d7f5b905f#diff-3849a1b9cacfa664dfb332b2993fda6bR95
To check where to navigate user - to contacts list or to this screen explaining our need of accessing user's address book